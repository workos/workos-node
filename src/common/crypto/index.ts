export * from './NodeCryptoProvider';
export * from './SubtleCryptoProvider';
export * from './CryptoProvider';
