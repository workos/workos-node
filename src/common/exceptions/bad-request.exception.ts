import { RequestException } from '../interfaces/request-exception.interface';

export class BadRequestException extends Error implements RequestException {
  readonly status = 400;
  readonly name = 'BadRequestException';
  readonly message: string = 'Bad request';
  readonly code?: string;
  readonly errors?: unknown[];
  readonly requestID: string;

  constructor({
    code,
    errors,
    message,
    requestID,
  }: {
    code?: string;
    errors?: unknown[];
    message?: string;
    requestID: string;
  }) {
    super();

    this.requestID = requestID;

    if (message) {
      this.message = message;
    }

    if (code) {
      this.code = code;
    }

    if (errors) {
      this.errors = errors;
    }
  }
}
