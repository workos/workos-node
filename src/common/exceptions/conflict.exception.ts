import { RequestException } from '../interfaces/request-exception.interface';

export class ConflictException extends Error implements RequestException {
  readonly status = 409;
  readonly name = 'ConflictException';
  readonly requestID: string;

  constructor({
    error,
    message,
    requestID,
  }: {
    error?: string;
    message?: string;
    requestID: string;
  }) {
    super();

    this.requestID = requestID;

    if (message) {
      this.message = message;
    } else if (error) {
      this.message = `Error: ${error}`;
    } else {
      this.message = `An conflict has occurred on the server.`;
    }
  }
}
