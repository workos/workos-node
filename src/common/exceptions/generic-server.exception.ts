import { HttpException } from '../interfaces';

export class GenericServerException implements HttpException {
  readonly name: string = 'GenericServerException';
  readonly message: string = 'The request could not be completed.';

  constructor(
    readonly status: number,
    message: string | undefined,
    readonly requestID: string,
  ) {
    if (message) {
      this.message = message;
    }
  }
}
