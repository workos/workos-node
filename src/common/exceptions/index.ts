export * from './generic-server.exception';
export * from './no-api-key-provided.exception';
export * from './not-found.exception';
export * from './oauth.exception';
export * from './signature-verification.exception';
export * from './unauthorized.exception';
export * from './unprocessable-entity.exception';
