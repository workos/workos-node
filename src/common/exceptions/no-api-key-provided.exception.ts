export class NoApiKeyProvidedException extends Error {
  readonly status = 500;
  readonly name = 'NoApiKeyProvidedException';
  readonly message =
    `Missing API key. Pass it to the constructor (new WorkOS("sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU")) ` +
    `or define it in the WORKOS_API_KEY environment variable.`;
}
