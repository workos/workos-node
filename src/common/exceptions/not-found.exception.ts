import { HttpException } from '../interfaces';

export class NotFoundException implements HttpException {
  readonly status: number = 404;
  readonly name: string = 'NotFoundException';
  readonly message: string;

  constructor(path: string, readonly requestID: string) {
    this.message = `The requested path '${path}' could not be found.`;
  }
}
