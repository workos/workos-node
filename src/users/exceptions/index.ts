import { UserResponse } from '../interfaces';

export class UnexpectedUserTypeException extends Error {
  constructor(user: UserResponse) {
    super();
    this.message = `Unsupported user type: ${user.user_type} received from API.`;
  }
}
