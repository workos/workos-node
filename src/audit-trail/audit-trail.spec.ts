import axios from 'axios';
import MockAdapater from 'axios-mock-adapter';

import WorkOS from '../workos';
import { UnauthorizedException } from '../common/exceptions';

const mock = new MockAdapater(axios);
const event = {
  group: 'WorkOS',
  actor_name: 'WorkOS@example.com',
  actor_id: 'user_1',
  location: ' 192.0.0.8',
  occurred_at: new Date(0),
  target_name: 'Security Audit 2018',
  target_id: 'document_39127',
  action_type: 'U',
  action: 'document.updated',
};

describe('AuditTrail', () => {
  describe('createEvent', () => {
    describe('when the api responds with a 201 CREATED', () => {
      describe('with an idempotency key', () => {
        it('includes an idempotency key with request', async () => {
          mock.onPost().reply(201, { success: true });

          const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU');

          await expect(
            workos.auditTrail.createEvent(event, {
              idempotencyKey: 'the-idempotency-key',
            }),
          ).resolves.toBeUndefined();

          expect(mock.history.post[0].headers['Idempotency-Key']).toEqual(
            'the-idempotency-key',
          );
        });
      });

      it('posts Event successfully', async () => {
        mock.onPost().reply(201, { success: true });

        const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU');

        await expect(
          workos.auditTrail.createEvent(event),
        ).resolves.toBeUndefined();
      });
    });

    describe('when the api responds with a 401', () => {
      it('throws an UnauthorizedException', async () => {
        mock
          .onPost()
          .reply(
            401,
            { message: 'Unauthorized' },
            { 'X-Request-ID': 'a-request-id' },
          );

        const workos = new WorkOS('invalid apikey');

        await expect(
          workos.auditTrail.createEvent(event),
        ).rejects.toStrictEqual(new UnauthorizedException('a-request-id'));
      });
    });

    describe('when the api responds with a 422', () => {
      it('throws an UnprocessableEntity', async () => {
        const errors = [
          {
            field: 'target_id',
            code: 'target_id must be a string',
          },
          {
            field: 'occurred_at',
            code: 'occurred_at must be an ISO 8601 date string',
          },
        ];

        mock.onPost().reply(
          422,
          {
            message: 'Validation failed',
            errors,
          },
          { 'X-Request-ID': 'a-request-id' },
        );

        const workos = new WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU');

        await expect(
          workos.auditTrail.createEvent(event),
        ).rejects.toMatchSnapshot();
      });
    });
  });
});
