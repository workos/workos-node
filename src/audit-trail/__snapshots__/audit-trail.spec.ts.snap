// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuditTrail createEvent when the api responds with a 422 throws an UnprocessableEntity 1`] = `
UnprocessableEntityException {
  "message": "The following requirements must be met:
	target_id must be a string
	occurred_at must be an ISO 8601 date string
",
  "name": "UnprocessableEntityException",
  "requestID": "a-request-id",
  "status": 422,
}
`;
